src/main/kotlin/inspections/variable/VariableNamesInspection.kt

```diff
@@ -0,0 +1,81 @@
+package inspections.variable
+
+import com.intellij.codeInspection.*
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.project.Project
+import com.intellij.psi.JavaElementVisitor
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.PsiVariable
+import com.intellij.psi.SmartPsiElementPointer
+import com.intellij.refactoring.suggested.createSmartPointer
+import org.jetbrains.id.names.suggesting.IdNamesSuggestingService
+import org.jetbrains.id.names.suggesting.ModifiedMemberInplaceRenamer
+
+class VariableNamesInspection : AbstractBaseJavaLocalInspectionTool() {
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return VariableVisitor(holder)
+    }
+
+    class VariableVisitor(private val holder: ProblemsHolder) : JavaElementVisitor() {
+        private val probabilityCutoff: Double = 0.001
+
+        override fun visitVariable(variable: PsiVariable?) {
+            when {
+                variable == null -> return
+//                ProbabilitiesStorage.isIgnored(variable) -> return
+//                caretInsideVariable(variable) -> recalculateLater(variable)
+                else -> {
+                    // TODO: It might be removed but i think it would be better to store probabilities for each file and save only last 10 files, for example. For {@link inspections.method.MethodNamesInspection} it might be too slow to evaluate on every method every time.
```

Comment:
To decide on what cache strategy to use it's important to understand the lifetime and contracts on the visitor. Maybe it's triggered not so often and we don't need the cache. Or we could place the cache right inside the visitor and we don't care a lot about cache invalidation, etc. You can investigate this once we notice performance issues with the naive implementation.